### PR TITLE
HS-597: Push metrics to cortex, remove pushgateway

### PR DESCRIPTION
**Pull request by BenjaminJ**
## Description
The metrics processor service currently pushes metrics to Prometheus through pushgateway. It needs to change for BTO to push to Cortex.

Point at Cortex instead of pushgateway

Make any necessary changes to talk to Cortex's API

Must be tenant aware
## Jira link(s)
- https://issues.opennms.org/browse/HS-597

## Flagged for review
- is Instant.now()okay for metrics timestamp  
- not sure about the configuration
- branch jw/cortex is merged into this one

## Checklist
* [ ] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.

[HS-597]: https://opennms.atlassian.net/browse/HS-597?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: metrics-processor/pom.xml [hunk 1]
```diff
@@ -68,21 +68,31 @@
...
-	</dependencies>
+        <dependency>
+            <groupId>com.squareup.okhttp3</groupId>
+            <artifactId>okhttp</artifactId>
+            <version>${okhttp.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.dropwizard.metrics</groupId>
+            <artifactId>metrics-core</artifactId>
+            <version>${metrics-core.version}</version>
```
Comment (yangli2014) [inline metrics-processor/pom.xml:88]: There is a "<dropwizard.metrics.version>4.2.10</dropwizard.metrics.version>" in the parent-pom. we might use the same version.

File: metrics-processor/src/main/resources/application.yml [hunk 2]
```diff
@@ -10,4 +10,9 @@ spring:
 kafka:
   topics: "task-set.results"
 
-prometheus.pushgateway.url: localhost:21080
+cortex.write.url: http://cortex:9000/api/v1/push
```
Comment (yangli2014) [inline metrics-processor/src/main/resources/application.yml:13]: This need to be added as an environment variable in the metrics processor deployment file.

File: parent-pom/pom.xml [hunk 3]
```diff
@@ -113,6 +113,12 @@
         <jetty.version>9.4.43.v20210629</jetty.version>
         <cron-utils.version>9.1.6</cron-utils.version>
         <eos.version>1.2.0</eos.version>
+        <tss-tests.version>1.0.0</tss-tests.version>
+        <metrics-core.version>4.2.6</metrics-core.version>
+        <common.version>1.1.0</common.version>
```
Comment (yangli2014) [inline parent-pom/pom.xml:118]: Where it is used?

File: metrics-processor/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+ * we use the querier.
+ * Docs:
+ * - https://cortexmetrics.io/docs/api/
+ * - https://prometheus.io/docs/prometheus/latest/querying/api/
+ *
+ * @author jwhite
+ */
+@Slf4j
+public class CortexTSS {
+    private static final Logger LOG = LoggerFactory.getLogger(CortexTSS.class);
```
Comment (yangli2014) [inline metrics-processor/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java:73]: If there is @Slf4j we don't need to create a logger again.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java [hunk 5]
```diff
@@ -117,85 +118,87 @@ public void consume(@Payload byte[] data, @Headers Map<String, Object> headers)
...
-
-    private Gauge getGaugeFrom(String name, String description, String unit) {
-        return gauges.compute(name, (key, gauge) -> {
-            if (gauge != null) {
-                return gauge;
-            }
-            var builder = Gauge.build().name(name).labelNames(TSDataProcessor.MONITOR_METRICS_LABEL_NAMES);
+        prometheus.PrometheusTypes.TimeSeries.Builder builder = prometheus.PrometheusTypes.TimeSeries.newBuilder();
+
+        if (response.getMetricsMap().isEmpty()) {
```
Comment (yangli2014) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:127]: I believe the metrics map in the response is the extra metrics and the response_time_msec should always be there.  We might need input from Jesse or others.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java [hunk 6]
```diff
@@ -117,85 +118,87 @@ public void consume(@Payload byte[] data, @Headers Map<String, Object> headers)
...
+                    .setValue(sanitizeMetricName(METRICS_NAME_PREFIX_MONITOR + k))));
+        }
 
-            if (!Strings.isNullOrEmpty(description)) {
-                builder.help(description);
-            }
-            if (!Strings.isNullOrEmpty(unit)) {
-                builder.unit(unit);
-            }
+        for (int i = 0; i < MONITOR_METRICS_LABEL_NAMES.length; i++) {
```
Comment (yangli2014) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:138]: We should remove the node_id label for minion metrics since it always is 0.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java [hunk 7]
```diff
@@ -117,85 +118,87 @@ public void consume(@Payload byte[] data, @Headers Map<String, Object> headers)
...
             response.getMonitorType().name(), String.valueOf(response.getNodeId())};
+        prometheus.PrometheusTypes.TimeSeries.Builder builder = prometheus.PrometheusTypes.TimeSeries.newBuilder();
         if (response.hasResult() && response.getMonitorType().equals(MonitorType.SNMP)) {
             Any collectorMetric = response.getResult();
             try {
                 var snmpResponse = collectorMetric.unpack(SnmpResponseMetric.class);
                 snmpResponse.getResultsList().forEach((snmpResult) -> {
-                    String metricName = snmpResult.getAlias();
-                    String description = metricName + " with oid " + snmpResult.getBase();
+                    builder.addLabels(PrometheusTypes.Label.newBuilder()
```
Comment (yangli2014) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:161]: I am not sure if this is correct, we might need a TimeSeries.Builder for teach snmpResult comparing the Promethuares version. It should create different metrics names based on each snmpResult.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java [hunk 8]
```diff
@@ -117,85 +118,87 @@ public void consume(@Payload byte[] data, @Headers Map<String, Object> headers)
         }
     }
 
-    private void processMonitorResponse(String tenantId, TaskResult result) {
+    private void processMonitorResponse(String tenantId, TaskResult result) throws IOException {
         MonitorResponse response = result.getMonitorResponse();
         String[] labelValues = {tenantId, response.getIpAddress(), result.getLocation(), result.getSystemId(), response.getMonitorType().name(), String.valueOf(response.getNodeId())};
```
Comment (yangli2014) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:123]: I am not sure if putting tenant id as a label in context is correct or not. According to the document "All Cortex components take the tenant ID from a header X-Scope-OrgID"

File: charts/opennms/templates/prometheus/prometheus-configmap.yaml [hunk 9]
```diff
@@ -21,3 +21,5 @@ data:
           - targets: ['{{ .Values.Prometheus.PushGateway.ServiceName }}:{{ .Values.Prometheus.PushGateway.Port }}']
             labels:
               pushgateway_instance: horizon-core-pushgateway
+    remote_write:
+      - url: http://{{ .Values.Cortex.ServiceName }}:{{ .Values.Cortex.Port }}/api/v1/push
```
Comment (yangli2014) [inline charts/opennms/templates/prometheus/prometheus-configmap.yaml:25]: What is the purpose of this config entry?

**Comment by yangli2014** [comment]
Do we still need Prometheus and Prometheus push gateway? If we don't need them we can remove them in another Jira ticket.

File: charts/opennms/templates/prometheus/prometheus-configmap.yaml (hunk 9 above)
Comment (BenjaminJ) [reply charts/opennms/templates/prometheus/prometheus-configmap.yaml:25]: It's from this PR https://github.com/OpenNMS/horizon-stream/pull/513

maybe we don't need this config anymore.

File: metrics-processor/pom.xml (hunk 1 above)
Comment (BenjaminJ) [reply metrics-processor/pom.xml]: fixed

File: metrics-processor/src/main/resources/application.yml (hunk 2 above)
Comment (BenjaminJ) [reply metrics-processor/src/main/resources/application.yml:13]: fixed

File: parent-pom/pom.xml (hunk 3 above)
Comment (BenjaminJ) [reply parent-pom/pom.xml]: removed

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 6 above)
Comment (BenjaminJ) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:138]: I'm not sure to understdand why we don't need the node_id, if we add a device, it won't have the same node_id.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 8 above)
Comment (BenjaminJ) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:123]: I was not sure either so I put it in both places. I don't know yet how to test multi tenancy to see if it's needed as a label or not

**Comment by cgorantla** [comment]
Encountered this while persisting collector metrics:

```
2023-01-04 22:39:55.932  INFO 1 --- [onPool-worker-3] o.o.horizon.tsdata.TSDataProcessor       : Have collector response, tenant-id: opennms-prime; task-id=ip=192.168.1.205/snmp-collector;
level=error ts=2023-01-04T22:39:55.940067159Z caller=push.go:51 org_id=fake msg="push error" err="rpc error: code = Code(400) desc = series has too many labels (actual: 32, limit: 30) series: 'ifHighSpeed{instance=\"192.168.1.205\", location=\"Default\", monitor=\"SNMP\", node_id=\"3\", system_id=\"opennms-minion-68d685f799-ch9fz\", tenant_id=\"opennms-prime\"}'"
2023-01-04 22:39:55.943 ERROR 1 --- [cortex:9000/...] org.opennms.timeseries.cortex.CortexTSS  : Error occurred while storing result, sample will be lost.

java.lang.RuntimeException: Writing to Prometheus failed: 400 - Bad Request: series has too many labels (actual: 32, limit: 30) series: 'ifHighSpeed{instance="192.168.1.205", location="Default", monitor="SNMP", node_id="3", system_id="opennms-minion-68d685f799-ch9fz", tenant_id="opennms-prime"}'

	at org.opennms.timeseries.cortex.CortexTSS$1.onResponse(CortexTSS.java:164) ~[classes/:na]
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174) ~[okhttp-3.14.0.jar:na]
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32) ~[okhttp-3.14.0.jar:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
	at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
```

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 7 above)
Comment (cgorantla) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:161]: yes, each SnmpResult is a different metric which needs to be persisted with different `__name__` label

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 5 above)
Comment (cgorantla) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:127]: yes, metrics maps contains more metrics apart from `response_time_msec`.  These are optional.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 6 above)
Comment (cgorantla) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:138]: Minions are not nodes, so they won't have node_id.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 8 above)
Comment (cgorantla) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:123]: we can remove `tenantId` label.  It's not needed.

**Review by cgorantla — COMMENTED** [review]
Except for the collector response failing to persist because of duplicate names being too many, this looks good to me. 

We can move this from draft PR to  actual PR.

Also verified that metrics are now assigned to time range. Yay!

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java (hunk 6 above)
Comment (yangli2014) [reply metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:138]: The minion metrics type is ECHO

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java [hunk 10]
```diff
@@ -117,85 +118,89 @@ public void consume(@Payload byte[] data, @Headers Map<String, Object> headers)
...
-    private void processMonitorResponse(String tenantId, TaskResult result) {
+    private void processMonitorResponse(String tenantId, TaskResult result) throws IOException {
         MonitorResponse response = result.getMonitorResponse();
-        String[] labelValues = {tenantId, response.getIpAddress(), result.getLocation(), result.getSystemId(), response.getMonitorType().name(), String.valueOf(response.getNodeId())};
-        Gauge gauge = getGaugeFrom(METRICS_NAME_RESPONSE, "Monitor round trip response time", METRICS_UNIT_MS);
-        gauge.labels(labelValues).set(response.getResponseTimeMs());
-        Map<String, String> labels = new HashMap<>();
-        for (int i = 0; i < MONITOR_METRICS_LABEL_NAMES.length; i++) {
-            labels.put(MONITOR_METRICS_LABEL_NAMES[i], labelValues[i]);
+        String[] labelValues = {response.getIpAddress(), result.getLocation(), result.getSystemId(), response.getMonitorType().name(), String.valueOf(response.getNodeId())};
```
Comment (cgorantla) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:123]: Beginning to think we should create proper mapping for these instead of using arrays for names and values.  May be cleanup this in another PR

File: charts/opennms/templates/prometheus/cortex-configmap.yaml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  namespace: {{ .Release.Namespace }}
+data:
+  cortex.yml: |
+    # Configuration for running Cortex in single-process mode.
+    # This should not be used in production.  It is only for getting started
+    # and development.
+    
+    # Disable the requirement that every request to Cortex has a
+    # X-Scope-OrgID header. `fake` will be substituted in instead.
+    #auth_enabled: false
```
Comment (cgorantla) [inline charts/opennms/templates/prometheus/cortex-configmap.yaml:16]: nit: Remove these lines

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/ApplicationConfig.java [hunk 12]
```diff
@@ -28,19 +28,34 @@
...
+import org.opennms.timeseries.cortex.CortexTSSConfig;
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
 @Configuration
 public class ApplicationConfig {
-    @Value("${prometheus.pushgateway.url}")
-    private String dataPushURL;
+    @Value("${cortex.write.url}")
```
Comment (cgorantla) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/ApplicationConfig.java:39]: nit: good to have line breaks for all these.

File: metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java [hunk 13]
```diff
@@ -117,85 +118,98 @@ public void consume(@Payload byte[] data, @Headers Map<String, Object> headers)
...
+                            builder.addSamples(PrometheusTypes.Sample.newBuilder()
+                                .setTimestamp(now)
+                                .setValue(metric));
                             break;
                     }
-
-                });
+                    cortexTSS.store(tenantId, builder);
+                }
             } catch (InvalidProtocolBufferException e) {
```
Comment (cgorantla) [inline metrics-processor/src/main/java/org/opennms/horizon/tsdata/TSDataProcessor.java:209]: I see that if an exception happens for a metric, other subsequent metrics are skipped, this catch can be moved to generic Exception.

**Review by cgorantla — COMMENTED** [review]
few comments to take care otherwise looks good to me.

Nice work @BenjaminJ !

File: charts/opennms/templates/prometheus/prometheus-configmap.yaml (hunk 9 above)
Comment (geraldhumphries) [reply charts/opennms/templates/prometheus/prometheus-configmap.yaml:25]: This configures Prometheus to push metrics to Cortex, just a way to get some data into it. We won't need it with this PR since we will be pushing directly to Cortex. @BenjaminJ will Prometheus end up totally unused after this PR is merged?

Comment (BenjaminJ) [reply charts/opennms/templates/prometheus/prometheus-configmap.yaml:25]: @geraldhumphries yes, I could remove prometheus and pushgateway from the tilt deployment, 
should I do it in this PR ?

**Review by cgorantla — APPROVED** [review]
LGTM!
